pkg/pubsub/pubsub_client.go

```diff
@@ -0,0 +1,123 @@
+package pubsub
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"strings"
+
+	"github.com/open-policy-agent/gatekeeper/pkg/pubsub/dapr"
+	"github.com/open-policy-agent/gatekeeper/pkg/pubsub/rabbitmq"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func init() {
+	flag.Var(pubSubs, "pub-sub-tool", "Preferred pub-sub tool, e.g dapr, rabbitmq. This flag can be declared more than once. Omitting will default to not using any pub sub tool.")
+}
+
+var pubSubs = newPubSubSet(map[string]PubSub{
+	dapr.Name:     &dapr.Dapr{},
+	rabbitmq.Name: &rabbitmq.RabbitMQ{},
+},
+)
+
+// PubSub is the interface that wraps pubsub methods.
+type PubSub interface {
+	// Publish single message over a specific topic/channel
+	Publish(data interface{}, topic string) error
+
+	// Publish a batch of messages over a specific topic/channel
+	PublishBatch(data interface{}, topic string) error
+
+	// Initiate the pub sub client
+	NewClient(ctx context.Context, client client.Client) error
+
+	// Get the name of pub sub tool
+	GetName() string
+
+	// Determines if publish batching is enabled or not
+	IsBatchingEnabled() bool
+}
+
+type pubSubSet struct {
+	supportedPubSub map[string]PubSub
+	enabledPubSub   map[string]PubSub
+}
+
+func newPubSubSet(pubSubs map[string]PubSub) *pubSubSet {
+	supported := make(map[string]PubSub)
+	enabled := make(map[string]PubSub)
+	set := &pubSubSet{
+		supportedPubSub: supported,
+		enabledPubSub:   enabled,
+	}
+	for name := range pubSubs {
+		set.AddSupportedTool(name, pubSubs[name])
+	}
+	return set
+}
+
+func (ps *pubSubSet) String() string {
+	pubSubNames := make([]string, 0)
+	for name := range ps.supportedPubSub {
+		pubSubNames = append(pubSubNames, name)
+	}
+	return fmt.Sprintf("%s", pubSubNames)
+}
+
+func (ps *pubSubSet) Set(s string) error {
+	splt := strings.Split(s, ",")
+	for _, v := range splt {
+		lower := strings.ToLower(v)
+		new, ok := ps.supportedPubSub[lower]
+		if !ok {
+			return fmt.Errorf("PubSub tool %s is not supported", lower)
+		}
+		ps.enabledPubSub[lower] = new
+	}
+	return nil
+}
+
+func (ps *pubSubSet) AddSupportedTool(name string, new PubSub) {
+	if _, ok := ps.supportedPubSub[name]; ok {
+		panic(fmt.Sprintf("pubsub %v registered twice", name))
+	}
+	ps.supportedPubSub[name] = new
+}
+
+func Tools() []PubSub {
+	if len(pubSubs.enabledPubSub) == 0 {
+		newDapr, ok := pubSubs.supportedPubSub[dapr.Name]
+		if !ok {
+			panic("dapr pubsub not registered, cannot use default pubsub")
+		}
+		return []PubSub{newDapr}
+	}
+	ret := make([]PubSub, 0, len(pubSubs.enabledPubSub))
+	for _, new := range pubSubs.enabledPubSub {
+		ret = append(ret, new)
+	}
+	return ret
+}
+
+func Publish(data interface{}, topic string) {
```

Comment:
First, do we want to have this topic name user configurable?

Additionally, Here we have two choices and based on that interface changes a little. Below are the two options we can go with that I can think of:
1. We have the same channel name across all tools for one purpose. In other words, there will be only one channel name, for example, `audit-pubsub` for all the tools for audit purposes. Not to say this will be a common shared channel across all the pub-sub tools, but different channels with the same name per tool. When we integrate pubsub with other parts of the gatekeeper, we will have other channel names such as `webhook-pubsub` for other purposes.  
2. We support multiple customizable channel names across all tools for one purpose. Meaning we allow users to define the channel name as `dapr-audit` for audit messages to be sent via Dapr, a channel name as `rabbitmq-audit` for audit messages sent via RabbitMQ, etc. This makes the interface a little more complex and I don't see much value from gk side. From the user's side, this can be useful as users can make use of existing channels if they have any, but again for that, we are assuming the user will have some of the pub-sub tools installed and integrated into their system.

Reply:
Taking a look at the high level makes this clear:

* We have one or more connections to a pubsub backend (say dapr or RabbitMQ)
* Each pubsub backend has its own set of topics
* we want to publish a message to a backend

That suggests a call format of `Publish(ctx, data, backend, topic)` (context is good to have for network requests to allow for things like timeouts).

IIRC we talked about Dapr being weird b/c you have only one Dapr connection. Their publish command is:

`client.PublishEvent(ctx, PUBSUB_NAME, TOPIC_NAME, []byte(strconv.Itoa(orderId))); `, taken from:

https://docs.dapr.io/developing-applications/building-blocks/pubsub/howto-publish-subscribe/

In that case, `backend` becomes `PUBSUB_NAME`

Because dapr is weird, we can fake multiple "connections" to dapr by creating a multiple virtual connections that call the same central connection, with different hardcoded values for PUBSUB_NAME.